# Comparison of MLE and GC results from SDS x CS82

After matching the SDStripe82 catalog with CS82 using both xmatching algorithms: `mle` and `gc`, we will now check how their results change -- if at all.

The catalogs are stored in the following files:
* gc: `swift_deepsky/xcat_sds_cs82_gc_7arcsec.csv`
* mle: `swift_deepsky/xcat_sds_cs82_mle_MAG_AUTO.csv`

In [1]:
gc_catalog = 'swift_deepsky/xcat_sds_cs82_gc_7arcsec.csv'
mle_catalog = 'swift_deepsky/xcat_sds_cs82_mle_MAG_AUTO.csv'

In [2]:
import pandas

df_gc = pandas.read_csv(gc_catalog, index_col=0, header=[0,1])
df_mle = pandas.read_csv(mle_catalog, index_col=0, header=[0,1])

In [3]:
df_gc

A                           B                          \
             RA       DEC OBJID ALPHA_J2000 DELTA_J2000       OBJID   
0     14.102000 -1.277310     0         NaN         NaN         NaN   
1     14.079733 -1.249499     1         NaN         NaN         NaN   
2     14.095850 -1.227643     2         NaN         NaN         NaN   
3     14.068408 -1.234838     3         NaN         NaN         NaN   
4     14.070508 -1.221223     4         NaN         NaN         NaN   
5     14.055663 -1.248697     5         NaN         NaN         NaN   
6     14.107788 -1.356168     6         NaN         NaN         NaN   
7     14.012692 -1.239376     7         NaN         NaN         NaN   
8     -3.830946 -0.202381     8         NaN         NaN         NaN   
9     -3.813942 -0.112661     9         NaN         NaN         NaN   
10    -3.913746 -0.116781    10         NaN         NaN         NaN   
11    -3.739321 -0.190447    11         NaN         NaN         NaN   
12    -3.740417 -0.200416    12         NaN         NaN         NaN   
13    -3.890392 -0.293661    13         NaN         NaN         NaN   
14    -3.756921 -0.026718    14   -3.756927   -0.025511   6203717.0   
15   -49.175729 -0.723092    15         NaN         NaN         NaN   
16   -49.187333 -0.677192    16         NaN         NaN         NaN   
17   -49.221463 -0.707921    17         NaN         NaN         NaN   
18   -49.068942 -0.761124    18         NaN         NaN         NaN   
19   -49.229983 -0.634637    19         NaN         NaN         NaN   
20   -49.072604 -0.793508    20         NaN         NaN         NaN   
21   -49.094679 -0.620873    21         NaN         NaN         NaN   
22   -49.057596 -0.651377    22         NaN         NaN         NaN   
23   -49.243667 -0.627425    23         NaN         NaN         NaN   
24   -49.275513 -0.791589    24         NaN         NaN         NaN   
25   -49.062721 -0.635699    25         NaN         NaN         NaN   
26   -49.228346 -0.593408    26         NaN         NaN         NaN   
27   -49.287579 -0.648607    27         NaN         NaN         NaN   
28   -49.176233 -0.574834    28         NaN         NaN         NaN   
29   -49.262371 -0.830387    29         NaN         NaN         NaN   
...         ...       ...   ...         ...         ...         ...   
2734  -4.524571 -1.516290  6346         NaN         NaN         NaN   
2735  -4.576487 -1.532624  6347         NaN         NaN         NaN   
2736  -4.449704 -1.489179  6348         NaN         NaN         NaN   
2737  -4.720821 -1.500833  6349         NaN         NaN         NaN   
2738  -4.770113 -1.367606  6350         NaN         NaN         NaN   
2739 -26.838925  0.504468  6373  -26.837278    0.504849  16471296.0   
2740 -26.824583  0.501955  6374  -26.824471    0.501865  16469443.0   
2741 -26.688596  0.569362  6382  -26.688659    0.569391  16483021.0   
2742 -26.679071  0.620482  6386  -26.679080    0.620619  16492857.0   
2743 -26.692963  0.638371  6387  -26.693149    0.638544  16495387.0   
2744 -26.633483  0.595959  6388  -26.633221    0.595969  16488182.0   
2745   8.158246 -1.010146  6409    8.159148   -1.009772  15982498.0   
2746 -27.305571 -0.224074  6466  -27.305257   -0.223879  16902006.0   
2747 -27.215000 -0.050653  6467  -27.215232   -0.050692  16924677.0   
2748 -27.338017 -0.181624  6468  -27.337565   -0.181652  16907516.0   
2749 -27.268325 -0.284991  6469  -27.267507   -0.284023  16894805.0   
2750 -27.187183 -0.009286  6470  -27.187598   -0.008953  16698105.0   
2751 -27.100696 -0.020542  6471  -27.100819   -0.021636  16385701.0   
2752 -27.017083 -0.178669  6472  -27.017797   -0.178864  16670320.0   
2753  29.729821  1.079033  6497         NaN         NaN         NaN   
2754  17.663104  0.366262  6624         NaN         NaN         NaN   
2755  17.668850  0.356361  6625         NaN         NaN         NaN   
2756  17.616296  0.455891  6628         NaN         NaN         NaN   
2757  17.611388  0.465086  6630   17.6

In [4]:
df_mle

A                                     B              \
                RA       DEC OBJID_A pos_err     OBJID_B ALPHA_J2000   
OBJID_A                                                                
0        14.102000 -1.277310     0.0     5.0         NaN         NaN   
1        14.079733 -1.249499     1.0     5.0         NaN         NaN   
2        14.095850 -1.227643     2.0     5.0         NaN         NaN   
3        14.068408 -1.234838     3.0     5.0         NaN         NaN   
4        14.070508 -1.221223     4.0     5.0         NaN         NaN   
5        14.055663 -1.248697     5.0     5.0         NaN         NaN   
6        14.107788 -1.356168     6.0     5.0         NaN         NaN   
7        14.012692 -1.239376     7.0     5.0         NaN         NaN   
8        -3.830946 -0.202381     8.0     5.0         NaN         NaN   
9        -3.813942 -0.112661     9.0     5.0         NaN         NaN   
10       -3.913746 -0.116781    10.0     5.0         NaN         NaN   
11       -3.739321 -0.190447    11.0     5.0         NaN         NaN   
12       -3.740417 -0.200416    12.0     5.0         NaN         NaN   
13       -3.890392 -0.293661    13.0     5.0         NaN         NaN   
14       -3.756921 -0.026718    14.0     5.0   6203632.0   -3.755398   
15      -49.175729 -0.723092    15.0     5.0         NaN         NaN   
16      -49.187333 -0.677192    16.0     5.0         NaN         NaN   
17      -49.221463 -0.707921    17.0     5.0         NaN         NaN   
18      -49.068942 -0.761124    18.0     5.0         NaN         NaN   
19      -49.229983 -0.634637    19.0     5.0         NaN         NaN   
20      -49.072604 -0.793508    20.0     5.0         NaN         NaN   
21      -49.094679 -0.620873    21.0     5.0         NaN         NaN   
22      -49.057596 -0.651377    22.0     5.0         NaN         NaN   
23      -49.243667 -0.627425    23.0     5.0         NaN         NaN   
24      -49.275513 -0.791589    24.0     5.0         NaN         NaN   
25      -49.062721 -0.635699    25.0     5.0         NaN         NaN   
26      -49.228346 -0.593408    26.0     5.0         NaN         NaN   
27      -49.287579 -0.648607    27.0     5.0         NaN         NaN   
28      -49.176233 -0.574834    28.0     5.0         NaN         NaN   
29      -49.262371 -0.830387    29.0     5.0         NaN         NaN   
...            ...       ...     ...     ...         ...         ...   
6469    -27.268325 -0.284991  6469.0     5.0  16894805.0  -27.267507   
6470    -27.187183 -0.009286  6470.0     5.0  16698429.0  -27.187237   
6471    -27.100696 -0.020542  6471.0     5.0  16385802.0  -27.099051   
6472    -27.017083 -0.178669  6472.0     5.0  16670320.0  -27.017797   
6497     29.729821  1.079033  6497.0     5.0         NaN         NaN   
6501     29.739083  1.160135  6501.0     5.0         NaN         NaN   
6624     17.663104  0.366262  6624.0     5.0         NaN         NaN   
6625     17.668850  0.356361  6625.0     5.0         NaN         NaN   
6626     17.678825  0.337546  6626.0     5.0         NaN         NaN   
6628     17.616296  0.455891  6628.0     5.0         NaN         NaN   
6630     17.611388  0.465086  6630.0     5.0   9448409.0   17.610906   
6631     17.606313  0.473746  6631.0     5.0         NaN         NaN   
6633     17.852179  0.445501  6633.0     5.0   9847831.0   17.851745   
6634     17.460071  0.393572  6634.0     5.0   9440078.0   17.458644   
6732     39.237083  0.726716  6732.0     5.0  14787476.0   39.236618   
6733     39.230850  0.613294  6733.0     5.0  14775413.0   39.232101   
6734     39.304346  0.671991  6734.0     5.0  14781889.0   39.305242   
6736     39.137842  0.527952  6736.0     5.0  14765002.0   39.137858   
6851     57.335662 -0.917533  6851.0     5.0         NaN         NaN   
6854     57.238637 -1.031111  6854.0     5.0         NaN         NaN   
6858     57.165358 -1.006882  6858.0     5.0         NaN         NaN   
6860     57.329054 -1.043098  6860.0     5.0         NaN 

## Let's check how many matching-pairs match between the catalogs

* This verification should gives us a measure on how much those two methods actually differ, and as such, an estimate of the xmatching error. 
* And it should point us to interesting sources or regions, where the nearest neighbor is *not* the "right" match.

In [5]:
objids_gc = df_gc[[('A','OBJID'),('B','OBJID')]]
objids_gc = objids_gc.dropna()
objids_gc = objids_gc.astype(int)
objids_gc.columns = ['OBJID_A','OBJID_B']
objids_gc.set_index('OBJID_A', inplace=True)

objids_gc.head()

OBJID_B
OBJID_A          
14        6203717
45       10331489
46       10331779
47       10336933
48       10349649

In [6]:
objids_mle = df_mle[[('A','OBJID_A'),('B','OBJID_B')]]
objids_mle = objids_mle.dropna()
objids_mle = objids_mle.astype(int)
objids_mle.columns = ['OBJID_A','OBJID_B']
objids_mle.set_index('OBJID_A', inplace=True)

objids_mle.head()

OBJID_B
OBJID_A          
14        6203632
45       10331489
46       10331970
47       10336874
48       10349673

In [7]:
all(objids_mle.index.sort_values() == objids_gc.index.sort_values())

True

In [8]:
dfx = objids_gc.join(objids_mle, how='outer', lsuffix='_gc', rsuffix='_mle')
dfx.head()

OBJID_B_gc  OBJID_B_mle
OBJID_A                         
14          6203717      6203632
45         10331489     10331489
46         10331779     10331970
47         10336933     10336874
48         10349649     10349673

In [9]:
df_diff = dfx['OBJID_B_gc'] == dfx['OBJID_B_mle']
df_diff.head()

OBJID_A
14    False
45     True
46    False
47    False
48    False
dtype: bool

In [10]:
equal_matches = df_diff.sum() / len(df_diff)

print('{:.1f}% of matches coincide between "gc" and "mle"'.format(equal_matches * 100))

70.4% of matches coincide between "gc" and "mle"


**70.4% of matches coincide between "gc" and "mle"**

## Who are the non-matching objects from 'mle'?

Let's check some properties of those ~30% objects that matched using the likelihood estimator.

In [11]:
_idx = df_diff.loc[~df_diff].index
df_nonmatching = df_mle.set_index(('A','OBJID_A')).loc[_idx]
df_nonmatching.head()

A                             B                          \
                RA       DEC pos_err     OBJID_B ALPHA_J2000 DELTA_J2000   
OBJID_A                                                                    
14       -3.756921 -0.026718     5.0   6203632.0   -3.755398   -0.027458   
46       24.657625 -0.668684     5.0  10331970.0   24.656740   -0.668031   
47       24.819667 -0.621067     5.0  10336874.0   24.821171   -0.621196   
48       24.696104 -0.495326     5.0  10349673.0   24.695655   -0.495929   
52       24.863604 -0.518196     5.0  10347523.0   24.863112   -0.518832   

                   AB_MAG_AUTO                                        \
          MAG_AUTO Reliability        LR                  duplicates   
OBJID_A                                                                
14       19.368633    0.764507  0.019682             6203717;6203774   
46       21.149553    0.446835  0.012284           10331779;10331716   
47       21.617054    0.399341  0.007689  10336933;10336804;10336854   
48       20.810715    0.598017  0.016153  10349649;10349580;10349671   
52       22.349245    0.565719  0.007371                    10347561   

                                                            \
                                             duplicates_LR   
OBJID_A                                                      
14              0.0024608894214439452;0.003601914698632182   
46                0.00490188752174791;0.010305705467634313   
47       0.0025365640202913682;0.0037324721520687115;0....   
48       0.00479203354230448;0.0028684722636638463;0.00...   
52                                    0.005658521449655007   

                                                            
                                              duplicates_R  
OBJID_A                                                     
14                     0.09558660098397;0.1399066451650775  
46                  0.17830270990683272;0.3748627858612379  
47       0.131745698030017;0.19385954587316417;0.275053...  
48       0.177410883069921;0.10619671019945533;0.118375...  
52                                      0.4342814700655355

In [12]:
29.59 < 100*len(df_nonmatching)/len(df_diff) < 29.60

True

In [13]:
from bokeh.io import output_notebook

output_notebook()

Loading BokehJS ...

In [14]:
def plot_distros(data1, data2, legend1, legend2):
    from bokeh.plotting import figure

    p = figure(plot_width=600, plot_height=400)

    bins = (data1.min(), data1.max(), 50)

    p = histogram(p, data1.dropna(), color='yellow', bins=bins, legend=legend1)
    p = histogram(p, data2.dropna(), color='blue', bins=bins, legend=legend2)
    p.legend.location = "top_left"
    
    return p

In [15]:
# HISTOGRAM

def histogram(fig, data, color, bins, legend):
    import numpy as np
    
    bins = np.linspace(*bins)
    
    hist, edges = np.histogram(data, density=True, bins=bins)
    
    fig.quad(top=hist, bottom=0, left=edges[:-1], right=edges[1:], 
             fill_color=color, line_color="#033649", fill_alpha=0.5,
             legend=legend)
    
    return fig

## .. a bigger sample from CS82

In [17]:
df_cs82_reduced = pandas.read_csv('cs82_photoauto_reduced_for_sds.csv', index_col=0)
df_cs82_reduced.head()

OBJID_B  ALPHA_J2000  DELTA_J2000   MAG_AUTO
0   6203717    -3.756927    -0.025511  25.151508
1  10331489    24.693307    -0.672908  21.615351
2  10331779    24.656872    -0.669435  23.783859
3  10336933    24.819401    -0.620192  25.077038
4  10349649    24.696638    -0.495811  24.011000

In [22]:
from bokeh.io import show

_col = ('B','MAG_AUTO')

data1 = df_cs82_reduced[_col[1]].dropna()
data2 = df_cs82_gc[_col[1]].dropna()
legend1 = 'background ({:d})'.format(len(data1))
legend2 = 'gc ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)

p.title.text = 'CS82 magnitude distribution of matched-vs-all sources'
p.title.align = 'center'
p.xaxis.axis_label = 'Magnitude (MAG_AUTO)'
p.yaxis.axis_label = '#'

show(p)

#### Plot: Background sample: 'MAG_AUTO' distribution of background sources (yellow) and counterparts (blue)

In [18]:
from bokeh.io import show

_col = ('B','MAG_AUTO')

data1 = df_cs82_reduced[_col[1]].dropna()
data2 = df_mle[_col].dropna()
legend1 = 'background ({:d})'.format(len(data1))
legend2 = 'mle ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)
p.title.text = 'CS82 magnitude distribution of matched-vs-all sources'
p.title.align = 'center'
p.xaxis.axis_label = 'Magnitude (MAG_AUTO)'
p.yaxis.axis_label = '#'

show(p)

#### Plot: Background sample: 'MAG_AUTO' distribution of background sources (yellow) and the gc-gc (blue) sub-sample

In [20]:
_idx = df_diff.loc[df_diff].index
df_matching = df_mle.set_index(('A','OBJID_A')).loc[_idx]
df_matching.head()

from bokeh.io import show

_col = ('B','MAG_AUTO')

data1 = df_cs82_reduced[_col[1]].dropna()
data2 = df_matching[_col].dropna()
legend1 = 'background ({:d})'.format(len(data1))
legend2 = 'gc & gc ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)

p.title.text = 'CS82 magnitude distribution of matched-vs-all sources'
p.title.align = 'center'
p.xaxis.axis_label = 'Magnitude (MAG_AUTO)'
p.yaxis.axis_label = '#'

show(p)

#### Plot: Background sample: 'MAG_AUTO' distribution of background sources (yellow) and the mle-gc (blue) sub-sample

In [19]:
from bokeh.io import show

_col = ('B','MAG_AUTO')

data1 = df_cs82_reduced[_col[1]].dropna()
data2 = df_nonmatching[_col].dropna()
legend1 = 'background ({:d})'.format(len(data1))
legend2 = 'mle-gc ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)

p.title.text = 'CS82 magnitude distribution of matched-vs-all sources'
p.title.align = 'center'
p.xaxis.axis_label = 'Magnitude (MAG_AUTO)'
p.yaxis.axis_label = '#'

show(p)

#### Plot: Cross-matched sample: 'MAG_AUTO' distribution of all sources (yellow) and the mle-only (blue) sub-sample

In [39]:
from bokeh.io import show

_col = ('B','MAG_AUTO')

data1 = df_mle[_col]
data2 = df_nonmatching[_col]
legend1 = 'all matched'
legend2 = 'mle only (mle-gc)'
p = plot_distros(data1, data2, legend1, legend2)

show(p)

## The GC catalog

In [21]:
df_cs82_gc = df_cs82_reduced.set_index('OBJID_B').loc[objids_gc['OBJID_B']]
df_cs82_gc.head()

ALPHA_J2000  DELTA_J2000   MAG_AUTO
OBJID_B                                      
6203717     -3.756927    -0.025511  25.151508
10331489    24.693307    -0.672908  21.615351
10331779    24.656872    -0.669435  23.783859
10336933    24.819401    -0.620192  25.077038
10349649    24.696638    -0.495811  24.011000

## Separation (GC) between pairs

In [23]:
def plot_distro(data):
    from bokeh.plotting import figure

    p = figure(plot_width=600, plot_height=400)

    bins = (data.min(), data.max(), 50)

    p = histogram(p, data.dropna(), color='blue', bins=bins, legend=None)
    
    return p

In [24]:
p = plot_distro(df_gc[('AB','separation')])
p.title.text = 'Distance separation between matching sources in SDS-x-CS82'
p.title.align = 'center'
p.xaxis.axis_label = 'Separation (arcsec)'
p.yaxis.axis_label = '#'

In [25]:
show(p)

In [26]:
df_gc[('AB','separation')].describe()

count    1105.000000
mean        2.901299
std         1.569761
min         0.052965
25%         1.700351
50%         2.643995
75%         3.864710
max         6.971090
Name: (AB, separation), dtype: float64

# Let's check the x-ray flux behaviour

In [27]:
sds_file = 'swift_deepsky/table_flux_detections_stripe82_unique.csv'

In [28]:
import pandas

df_sds = pandas.read_csv(sds_file, sep=';')
df_sds.head()

OBJID         RA       DEC            NH  ENERGY_SLOPE ENERGY_SLOPE_ERROR  \
0      0  14.102000 -1.277310  3.540000e+20           0.8          -999/-999   
1      1  14.079733 -1.249499  3.520000e+20           0.8          -999/-999   
2      2  14.095850 -1.227643  3.510000e+20           0.8          -999/-999   
3      3  14.068408 -1.234838  3.510000e+20           0.8          -999/-999   
4      4  14.070508 -1.221223  3.500000e+20           0.8          -999/-999   

   EXPOSURE_TIME  nufnu_3keV(erg.s-1.cm-2)  nufnu_error_3keV(erg.s-1.cm-2)  \
0         4572.8              7.969610e-14                    2.117700e-14   
1         4684.2              6.277090e-14                    1.975260e-14   
2         4649.6              5.501970e-14                    1.974560e-14   
3         4642.0              8.173270e-14                    2.256640e-14   
4         4587.3              7.365840e-14                    2.115000e-14   

   nufnu_0.5keV(erg.s-1.cm-2)  nufnu_error_0.5keV(erg.s-1.cm-2)  \
0                2.247330e-14                      1.552740e-14   
1                1.277270e-14                      1.252210e-14   
2                2.238590e-14                      1.751890e-14   
3                1.151060e-14                      1.187720e-14   
4                5.616880e-14                      2.418840e-14   

   upper_limit_0.5keV(erg.s-1.cm-2)  nufnu_1.5keV(erg.s-1.cm-2)  \
0                            -999.0                1.088260e-13   
1                            -999.0                7.075490e-14   
2                            -999.0                4.655630e-14   
3                            -999.0                9.568910e-14   
4                            -999.0                4.673000e-14   

   nufnu_error_1.5keV(erg.s-1.cm-2)  upper_limit_1.5keV(erg.s-1.cm-2)  \
0                      4.027270e-14                            -999.0   
1                      3.430380e-14                            -999.0   
2                      3.005940e-14                            -999.0   
3                      3.936350e-14                            -999.0   
4                      2.649940e-14                            -999.0   

   nufnu_4.5keV(erg.s-1.cm-2)  nufnu_error_4.5keV(erg.s-1.cm-2)  \
0                9.370070e-14                      4.601960e-14   
1                1.066430e-13                      5.170340e-14   
2                9.353370e-14                      5.227960e-14   
3                1.442380e-13                      5.933510e-14   
4                9.390950e-14                      4.601620e-14   

   upper_limit_4.5keV(erg.s-1.cm-2)  
0                            -999.0  
1                            -999.0  
2                            -999.0  
3                            -999.0  
4                            -999.0

In [29]:
df_sds.set_index('OBJID', inplace=True)
df_sds.head()

RA       DEC            NH  ENERGY_SLOPE ENERGY_SLOPE_ERROR  \
OBJID                                                                       
0      14.102000 -1.277310  3.540000e+20           0.8          -999/-999   
1      14.079733 -1.249499  3.520000e+20           0.8          -999/-999   
2      14.095850 -1.227643  3.510000e+20           0.8          -999/-999   
3      14.068408 -1.234838  3.510000e+20           0.8          -999/-999   
4      14.070508 -1.221223  3.500000e+20           0.8          -999/-999   

       EXPOSURE_TIME  nufnu_3keV(erg.s-1.cm-2)  \
OBJID                                            
0             4572.8              7.969610e-14   
1             4684.2              6.277090e-14   
2             4649.6              5.501970e-14   
3             4642.0              8.173270e-14   
4             4587.3              7.365840e-14   

       nufnu_error_3keV(erg.s-1.cm-2)  nufnu_0.5keV(erg.s-1.cm-2)  \
OBJID                                                               
0                        2.117700e-14                2.247330e-14   
1                        1.975260e-14                1.277270e-14   
2                        1.974560e-14                2.238590e-14   
3                        2.256640e-14                1.151060e-14   
4                        2.115000e-14                5.616880e-14   

       nufnu_error_0.5keV(erg.s-1.cm-2)  upper_limit_0.5keV(erg.s-1.cm-2)  \
OBJID                                                                       
0                          1.552740e-14                            -999.0   
1                          1.252210e-14                            -999.0   
2                          1.751890e-14                            -999.0   
3                          1.187720e-14                            -999.0   
4                          2.418840e-14                            -999.0   

       nufnu_1.5keV(erg.s-1.cm-2)  nufnu_error_1.5keV(erg.s-1.cm-2)  \
OBJID                                                                 
0                    1.088260e-13                      4.027270e-14   
1                    7.075490e-14                      3.430380e-14   
2                    4.655630e-14                      3.005940e-14   
3                    9.568910e-14                      3.936350e-14   
4                    4.673000e-14                      2.649940e-14   

       upper_limit_1.5keV(erg.s-1.cm-2)  nufnu_4.5keV(erg.s-1.cm-2)  \
OBJID                                                                 
0                                -999.0                9.370070e-14   
1                                -999.0                1.066430e-13   
2                                -999.0                9.353370e-14   
3                                -999.0                1.442380e-13   
4                                -999.0                9.390950e-14   

       nufnu_error_4.5keV(erg.s-1.cm-2)  upper_limit_4.5keV(erg.s-1.cm-2)  
OBJID                                                                      
0                          4.601960e-14                            -999.0  
1                          5.170340e-14                            -999.0  
2                          5.227960e-14                            -999.0  
3                          5.933510e-14                            -999.0  
4                          4.601620e-14                            -999.0

In [30]:
df_sds_matched = df_sds.loc[objids_gc.index]
df_sds_matched.head()

RA       DEC            NH  ENERGY_SLOPE ENERGY_SLOPE_ERROR  \
OBJID_A                                                                       
14       -3.756921 -0.026718  3.320000e+20           0.8          -999/-999   
45       24.693637 -0.672514  2.650000e+20           0.8          -999/-999   
46       24.657625 -0.668684  2.650000e+20           0.8          -999/-999   
47       24.819667 -0.621067  2.670000e+20           0.8          -999/-999   
48       24.696104 -0.495326  2.620000e+20           0.8          -999/-999   

         EXPOSURE_TIME  nufnu_3keV(erg.s-1.cm-2)  \
OBJID_A                                            
14              4826.6              5.258800e-14   
45             17176.0              9.711220e-15   
46             17138.0              1.439760e-14   
47             15367.0              1.321040e-14   
48             15398.6              2.047130e-14   

         nufnu_error_3keV(erg.s-1.cm-2)  nufnu_0.5keV(erg.s-1.cm-2)  \
OBJID_A                                                               
14                         1.401600e-14                7.640390e-14   
45                         3.698730e-15                4.282920e-15   
46                         4.383680e-15                4.616470e-15   
47                         4.249480e-15                9.999910e-15   
48                         5.336760e-15                1.202760e-14   

         nufnu_error_0.5keV(erg.s-1.cm-2)  upper_limit_0.5keV(erg.s-1.cm-2)  \
OBJID_A                                                                       
14                           2.363380e-14                            -999.0   
45                           3.262950e-15                            -999.0   
46                           3.287120e-15                            -999.0   
47                           5.003590e-15                            -999.0   
48                           5.394080e-15                            -999.0   

         nufnu_1.5keV(erg.s-1.cm-2)  nufnu_error_1.5keV(erg.s-1.cm-2)  \
OBJID_A                                                                 
14                     2.679660e-14                      1.657780e-14   
45                     9.418440e-15                      5.908220e-15   
46                     1.691710e-14                      7.650940e-15   
47                     1.951080e-14                      8.467530e-15   
48                     1.766150e-14                      7.920770e-15   

         upper_limit_1.5keV(erg.s-1.cm-2)  nufnu_4.5keV(erg.s-1.cm-2)  \
OBJID_A                                                                 
14                                 -999.0                8.094140e-14   
45                                 -999.0                1.432350e-14   
46                                 -999.0                2.058300e-14   
47                                 -999.0                1.887070e-14   
48                                 -999.0                2.686790e-14   

         nufnu_error_4.5keV(erg.s-1.cm-2)  upper_limit_4.5keV(erg.s-1.cm-2)  
OBJID_A                                                                      
14                          -5.387110e-11                      2.677380e-13  
45                           8.985180e-15                     -9.990000e+02  
46                           1.034270e-14                     -9.990000e+02  
47                          -1.668350e-11                      4.911330e-14  
48                           1.204960e-14                     -9.990000e+02

In [31]:
df_sds_nonmatched = df_sds.loc[~df_sds.index.isin(objids_gc.index)]
df_sds_nonmatched.head()

RA       DEC            NH  ENERGY_SLOPE ENERGY_SLOPE_ERROR  \
OBJID                                                                       
0      14.102000 -1.277310  3.540000e+20           0.8          -999/-999   
1      14.079733 -1.249499  3.520000e+20           0.8          -999/-999   
2      14.095850 -1.227643  3.510000e+20           0.8          -999/-999   
3      14.068408 -1.234838  3.510000e+20           0.8          -999/-999   
4      14.070508 -1.221223  3.500000e+20           0.8          -999/-999   

       EXPOSURE_TIME  nufnu_3keV(erg.s-1.cm-2)  \
OBJID                                            
0             4572.8              7.969610e-14   
1             4684.2              6.277090e-14   
2             4649.6              5.501970e-14   
3             4642.0              8.173270e-14   
4             4587.3              7.365840e-14   

       nufnu_error_3keV(erg.s-1.cm-2)  nufnu_0.5keV(erg.s-1.cm-2)  \
OBJID                                                               
0                        2.117700e-14                2.247330e-14   
1                        1.975260e-14                1.277270e-14   
2                        1.974560e-14                2.238590e-14   
3                        2.256640e-14                1.151060e-14   
4                        2.115000e-14                5.616880e-14   

       nufnu_error_0.5keV(erg.s-1.cm-2)  upper_limit_0.5keV(erg.s-1.cm-2)  \
OBJID                                                                       
0                          1.552740e-14                            -999.0   
1                          1.252210e-14                            -999.0   
2                          1.751890e-14                            -999.0   
3                          1.187720e-14                            -999.0   
4                          2.418840e-14                            -999.0   

       nufnu_1.5keV(erg.s-1.cm-2)  nufnu_error_1.5keV(erg.s-1.cm-2)  \
OBJID                                                                 
0                    1.088260e-13                      4.027270e-14   
1                    7.075490e-14                      3.430380e-14   
2                    4.655630e-14                      3.005940e-14   
3                    9.568910e-14                      3.936350e-14   
4                    4.673000e-14                      2.649940e-14   

       upper_limit_1.5keV(erg.s-1.cm-2)  nufnu_4.5keV(erg.s-1.cm-2)  \
OBJID                                                                 
0                                -999.0                9.370070e-14   
1                                -999.0                1.066430e-13   
2                                -999.0                9.353370e-14   
3                                -999.0                1.442380e-13   
4                                -999.0                9.390950e-14   

       nufnu_error_4.5keV(erg.s-1.cm-2)  upper_limit_4.5keV(erg.s-1.cm-2)  
OBJID                                                                      
0                          4.601960e-14                            -999.0  
1                          5.170340e-14                            -999.0  
2                          5.227960e-14                            -999.0  
3                          5.933510e-14                            -999.0  
4                          4.601620e-14                            -999.0

In [32]:
print(len(df_sds_matched))
print(len(df_sds_nonmatched))

1105
1659


In [33]:
from bokeh.io import show
from numpy import log10 as log

_col = 'nufnu_3keV(erg.s-1.cm-2)'

data1 = df_sds_nonmatched[_col].dropna()
data2 = df_sds_matched[_col].dropna()
                          
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

p = plot_distros(log(data1), log(data2), legend1, legend2)

_cnm = _col.split('(')[0]
p.title.text = 'SDS x-ray ({}) flux distribution'.format(_cnm)
p.title.align = 'center'
p.xaxis.axis_label = 'log( nuFnu_3keV [erg.s-1.cm-2] )'
p.yaxis.axis_label = '#'

p.legend.location = 'top_right'

show(p)

In [34]:
from bokeh.io import show
from numpy import log10 as log

_col = 'NH'

data1 = df_sds_nonmatched[_col].dropna()
data2 = df_sds_matched[_col].dropna()
                          
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)

p.title.text = 'SDS NH distribution'
p.title.align = 'center'
p.xaxis.axis_label = '{}'.format(_col)
p.yaxis.axis_label = '#'

p.legend.location = 'top_right'

show(p)

In [35]:
from bokeh.io import show
from numpy import log10 as log

_col = 'EXPOSURE_TIME'

data1 = df_sds_nonmatched[_col].dropna()
data2 = df_sds_matched[_col].dropna()
                          
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

p = plot_distros(data1, data2, legend1, legend2)

p.title.text = 'SDS exposure-time distribution'
p.title.align = 'center'
p.xaxis.axis_label = '{}'.format(_col)
p.yaxis.axis_label = '#'

p.legend.location = 'top_right'

show(p)

In [36]:
from bokeh.io import show
from numpy import log10 as log

_col = 'nufnu_0.5keV(erg.s-1.cm-2)'

data1 = df_sds_nonmatched[_col].dropna()
data1 = data1.loc[data1>0]
data2 = df_sds_matched[_col].dropna()
data2 = data2.loc[data2>0]
                          
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

p = plot_distros(log(data1), log(data2), legend1, legend2)

_cnm = _col.split('(')[0]
p.title.text = 'SDS x-ray ({}) flux distribution'.format(_cnm)
p.title.align = 'center'
p.xaxis.axis_label = 'log( {} )'.format(_col)
p.yaxis.axis_label = '#'

p.legend.location = 'top_right'

show(p)

In [37]:
from bokeh.io import show
from numpy import log10 as log

_col = 'nufnu_1.5keV(erg.s-1.cm-2)'

data1 = df_sds_nonmatched[_col].dropna()
data1 = data1.loc[data1>0]
data2 = df_sds_matched[_col].dropna()
data2 = data2.loc[data2>0]
                          
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

p = plot_distros(log(data1), log(data2), legend1, legend2)

_cnm = _col.split('(')[0]
p.title.text = 'SDS x-ray ({}) flux distribution'.format(_cnm)
p.title.align = 'center'
p.xaxis.axis_label = 'log( {} )'.format(_col)
p.yaxis.axis_label = '#'

p.legend.location = 'top_right'

show(p)

In [38]:
def plot_distro_flux(title, data1, data2, colname, legend1, legend2):
    from numpy import log10 as log

    _col = colname
    data1 = data1.loc[data1>0]
    data2 = data2.loc[data2>0]

    p = plot_distros(log(data1), log(data2), legend1, legend2)

    p.title.text = title
    p.title.align = 'center'
    p.xaxis.axis_label = 'log( {} )'.format(_col)
    p.yaxis.axis_label = '#'

    p.legend.location = 'top_right'

    show(p)

_col = 'nufnu_4.5keV(erg.s-1.cm-2)'
_cnm = _col.split('(')[0]
title = 'SDS x-ray ({}) flux distribution'.format(_cnm)

data1 = df_sds_nonmatched[_col].dropna()
data2 = df_sds_matched[_col].dropna()
legend1 = 'non-matched ({:d})'.format(len(data1))
legend2 = 'matched ({:d})'.format(len(data2))

from bokeh.io import show
p = plot_distro_flux(title, data1, data2, _col, legend1, legend2)